### PR TITLE
Router: Optimize catch-all routes

### PR DESCRIPTION
**Pull request by iquito**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | unsure - not quite finished yet
| License       | MIT

The typical catch-all route (example: `@Route("/{url}", requirements={"url" = ".*"})`) still leads to a preg_match being executed, even though the route will always match. When using Symfony with a catch-all route to "pass along" the request to another system or part of the application this preg_match gets executed on every request without any added benefit.

I attempted to improve this with the changes in this pull request (condition "true" instead of the regex), although it still has flaws:

- The named catch-all parameter (for example "url" in `@Route("/{url}", requirements={"url" = ".*"})`) is empty instead of containing the whole URL because I was not sure where to get the URL or how to add it to the Dumper. I thought it more sensible to just post a partial solution instead of no solution or going through the Router code for a lot longer.
- The regex to find a "catch-all" route is a bit crude, and I was not sure if `[A-Za-z0-9_]+` actually contains all allowed characters for placeholders in routes.
- Tests probably need to be modified too, but I thought it better to just post something small to see if this is even a feature which Symfony wants to implement in this manner.

Hope this still helps, and would be glad to improve it further after some feedback!

### CODE REVIEW

**Comment by stof** [comment]
having no change in tests look suspicious to me

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 1]
```diff
@@ -244,9 +244,14 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
                 $regex = substr($regex, 0, $pos).'/?$'.substr($regex, $pos + 2);
                 $hasTrailingSlash = true;
             }
-            $conditions[] = sprintf('preg_match(%s, $pathinfo, $matches)', var_export($regex, true));
 
-            $matches = true;
+            if (!preg_match('#^'.preg_quote('#^/(?P<', '#').'([A-Za-z0-9_]+)'.preg_quote('>.*)?$#s', '#').'$#', $regex)) {
```
Comment (stof) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: why building a regex dynamically when it does not involve any variable ? This looks totally unreadable (and useless) to me.

**Comment by stof** [comment]
> The named catch-all parameter (for example "url" in @Route("/{url}", requirements={"url" = ".*"})) is empty instead of containing the whole URL because I was not sure where to get the URL or how to add it to the Dumper. I thought it more sensible to just post a partial solution instead of no solution or going through the Router code for a lot longer.

When, the URL is the stored in the variable being matched by the regex. It is there in the matcher.
It cannot be know in the dumper of course, as the URL can be anything in the case of a catch-all (this is the point of a catch-all).

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php (hunk 1 above)
Comment (iquito) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: This is more like a proof-of-concept, as I have mentioned - I just looked at the regex which the Router creates for the catch-all route and matched it, and it works. An example of the regex with the placeholder "url" would be `#^/(?P<url>.*)?$#s`, so it looks for `#^/(?P<` then a placeholder variable name and then `>.*)?$#s`.

**Comment by iquito** [comment]
Yes of course, but the Dumper can create the code to replace the placeholder variable with the whole URL.

Because I am basically new to the Routing component and its source code and this is probably trivial to solve for somebody who knows it, but would only lead to a not-so-perfect solution by me, I thought I would just mention what is missing instead of trying a solution myself.

**Comment by stof** [comment]
@iquito my advice to see how you could optimize it would be to first keep the dumper out of the process: define the routes you need, then dump the matcher with the existing dumper and get the generated code (which is in your ``var/cache/dev`` folder). Then, start by editing this code by hand to build the matching code you need (keeping track of the diff with the initial code). Then, it will be easier to adapt the dumper code, as you will know what you need to dump

**Comment by iquito** [comment]
That is what I did - with my changes the dumped code is changed from

    // legacy_catchall_index
    if (preg_match('#^/(?P<url>.*)?$#s', $pathinfo, $matches)) {
        return $this->mergeDefaults(array_replace($matches, array('_route' => 'legacy_catchall_index')), array (  'url' => '',  '_controller' => 'LegacyBundle\\Controller\\CatchAllController::indexAction',));
    }

to this:

    // legacy_catchall_index
    if (true) {
        return array (  'url' => '',  '_controller' => 'LegacyBundle\\Controller\\CatchAllController::indexAction',  '_route' => 'legacy_catchall_index',);
    }

So the 'url' part is incorrect after the change, but it detects catch-all routes and replaces them with true instead pf the preg_match. because a catch-all route is basically a static route. That is what I tried to demonstrate with my pull request.

**Comment by stof** [comment]
well, the need to build the URL is precisely the part where I tell you to edit the matcher code directly, to see how you would write the logic. then, you can work on making the dumper generates this code.

**Comment by iquito** [comment]
Maybe we are not talking about the same thing - the one catch-all parameter would just get the whole URL, so probably `$pathinfo` in the dumper generated code, so the code which should be generated in my example would be:

    // legacy_catchall_index
    if (true) {
        return array (  'url' => $pathinfo,  '_controller' => 'LegacyBundle\\Controller\\CatchAllController::indexAction',  '_route' => 'legacy_catchall_index',);
    }

Then it would work exactly as intended. Because I don't know all the in-and-outs of the Router component, I just suggested a working solution which needs at least the above addition to be completely backwards-compatible.

I did not want to be frustrating, but I thought some code which almost works is better than just opening an issue with nothing at all to show for.

**Comment by nicolas-grekas** [comment]
@iquito unless you're able to provide a working implementation (requirements are still regexp so they need a PCRE check), I really doubt anybody will dive into this for you. There is not bug here, and the performance enhancement you suspect is unproven yet (only code would prove it.) I suggest to close this PR, unless you'd like to work on it and get something merged.

**Comment by fabpot** [comment]
Closing for now.

**Pull request closed without merging**